lib/ElmJson.js

```diff
@@ -0,0 +1,193 @@
+// @flow
+
+const fs = require('fs');
+const path = require('path');
+
+// Poor man’s type alias. We can’t use /*:: type Dependencies = ... */ because of:
+// https://github.com/prettier/prettier/issues/2597
+const Dependencies /*: { [string]: string } */ = {};
+
+const DirectAndIndirectDependencies /*: {
+  direct: typeof Dependencies,
+  indirect: typeof Dependencies,
+} */ = { direct: {}, indirect: {} };
+
+const ElmJson /*:
+  | {
+      type: 'application',
+      'source-directories': Array<string>,
+      dependencies: typeof DirectAndIndirectDependencies,
+      'test-dependencies': typeof DirectAndIndirectDependencies,
+      [string]: mixed,
+    }
+  | {
+      type: 'package',
+      dependencies: typeof Dependencies,
+      'test-dependencies': typeof Dependencies,
+      [string]: mixed,
+    } */ = {
+  type: 'package',
+  dependencies: Dependencies,
+  'test-dependencies': Dependencies,
+};
+
+function getPath(dir /*: string */) /*: string */ {
+  return path.join(dir, 'elm.json');
+}
+
+function write(dir /*: string */, elmJson /*: typeof ElmJson */) /*: void */ {
+  const elmJsonPath = getPath(dir);
+
+  try {
+    fs.writeFileSync(elmJsonPath, JSON.stringify(elmJson, null, 4) + '\n');
+  } catch (error) {
+    throw new Error(
+      `${elmJsonPath}\nFailed to write elm.json:\n${error.message}`
+    );
+  }
+}
+
+function read(dir /*: string */) /*: typeof ElmJson */ {
+  const elmJsonPath = getPath(dir);
+
+  try {
+    return readHelper(elmJsonPath);
+  } catch (error) {
+    throw new Error(
+      `${elmJsonPath}\nFailed to read elm.json:\n${error.message}`
+    );
+  }
+}
+
+function readHelper(elmJsonPath /*: string */) /*: typeof ElmJson */ {
+  const json = parseObject(
+    JSON.parse(fs.readFileSync(elmJsonPath, 'utf8')),
```

Comment:
Minor comment: `JSON.parse` can throw a `SyntaxError`

Reply:
Yes, I have tests for that. Everything in this file throws errors when something’s invalid.

Reply:
Maybe a comment at the top of the file noting that all these functions may throw would help here.

Reply:
Sure, I can add a comment, but isn’t this kind of obvious? The file is full of `throw new Error` and it feels like well-known stuff that reading files and parsing JSON throws errors if they fail?

Reply:
Do we not usually use a `Result` type now in the test runner? If someone saw our use of result types in some files they might expect us to handle all errors with a  `Result` type. I was just thinking it might help to clarify which paradigm we use in which file?

Reply:
I got rid of `Result` because it was much easier to deal with regular exceptions in the end. Especially in async code.

Reply:
Makes sense :+1: both results and exceptions have their benefits but consistency is most important